# JSON examples and exercise
****
+ get familiar with packages for dealing with JSON
+ study examples with JSON strings and files 
+ work on exercise to be completed and submitted 
****
+ reference: http://pandas.pydata.org/pandas-docs/stable/io.html#io-json-reader
+ data source: http://jsonstudio.com/resources/
****

In [1]:
import pandas as pd

## imports for Python, Pandas

In [2]:
import json
from pandas.io.json import json_normalize

## JSON example, with string

+ demonstrates creation of normalized dataframes (tables) from nested json string
+ source: http://pandas.pydata.org/pandas-docs/stable/io.html#normalization

In [3]:
# define json string
data = [{'state': 'Florida', 
         'shortname': 'FL',
         'info': {'governor': 'Rick Scott'},
         'counties': [{'name': 'Dade', 'population': 12345},
                      {'name': 'Broward', 'population': 40000},
                      {'name': 'Palm Beach', 'population': 60000}]},
        {'state': 'Ohio',
         'shortname': 'OH',
         'info': {'governor': 'John Kasich'},
         'counties': [{'name': 'Summit', 'population': 1234},
                      {'name': 'Cuyahoga', 'population': 1337}]}]

In [4]:
# use normalization to create tables from nested element
json_normalize(data, 'counties')

name  population
0        Dade       12345
1     Broward       40000
2  Palm Beach       60000
3      Summit        1234
4    Cuyahoga        1337

In [5]:
# further populate tables created from nested element
json_normalize(data, 'counties', ['state', 'shortname', ['info', 'governor']])

name  population    state shortname info.governor
0        Dade       12345  Florida        FL    Rick Scott
1     Broward       40000  Florida        FL    Rick Scott
2  Palm Beach       60000  Florida        FL    Rick Scott
3      Summit        1234     Ohio        OH   John Kasich
4    Cuyahoga        1337     Ohio        OH   John Kasich

****
## JSON example, with file

+ demonstrates reading in a json file as a string and as a table
+ uses small sample file containing data about projects funded by the World Bank 
+ data source: http://jsonstudio.com/resources/

In [8]:
# load json as string
json.load((open('data/world_bank_projects_less.json')))

[{'_id': {'$oid': '52b213b38594d8a2be17c780'},
  'approvalfy': 1999,
  'board_approval_month': 'November',
  'boardapprovaldate': '2013-11-12T00:00:00Z',
  'borrower': 'FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA',
  'closingdate': '2018-07-07T00:00:00Z',
  'country_namecode': 'Federal Democratic Republic of Ethiopia!$!ET',
  'countrycode': 'ET',
  'countryname': 'Federal Democratic Republic of Ethiopia',
  'countryshortname': 'Ethiopia',
  'docty': 'Project Information Document,Indigenous Peoples Plan,Project Information Document',
  'envassesmentcategorycode': 'C',
  'grantamt': 0,
  'ibrdcommamt': 0,
  'id': 'P129828',
  'idacommamt': 130000000,
  'impagency': 'MINISTRY OF EDUCATION',
  'lendinginstr': 'Investment Project Financing',
  'lendinginstrtype': 'IN',
  'lendprojectcost': 550000000,
  'majorsector_percent': [{'Name': 'Education', 'Percent': 46},
   {'Name': 'Education', 'Percent': 26},
   {'Name': 'Public Administration, Law, and Justice', 'Percent': 16},
   {'Name': 'Educatio

In [9]:
# load as Pandas dataframe
sample_json_df = pd.read_json('data/world_bank_projects_less.json')
sample_json_df

_id  approvalfy board_approval_month  \
0  {'$oid': '52b213b38594d8a2be17c780'}        1999             November   
1  {'$oid': '52b213b38594d8a2be17c781'}        2015             November   

      boardapprovaldate                                 borrower  \
0  2013-11-12T00:00:00Z  FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA   
1  2013-11-04T00:00:00Z                    GOVERNMENT OF TUNISIA   

            closingdate                              country_namecode  \
0  2018-07-07T00:00:00Z  Federal Democratic Republic of Ethiopia!$!ET   
1                   NaN                      Republic of Tunisia!$!TN   

  countrycode                              countryname countryshortname  \
0          ET  Federal Democratic Republic of Ethiopia         Ethiopia   
1          TN                      Republic of Tunisia          Tunisia   

                         ...                           sectorcode source  \
0                        ...                          ET,BS,ES,EP   IBRD   
1                        ...                                BZ,BS   IBRD   

   status  supplementprojectflg  \
0  Active                     N   
1  Active                     N   

                                              theme1  \
0      {'Name': 'Education for all', 'Percent': 100}   
1  {'Name': 'Other economic management', 'Percent...   

                                      theme_namecode themecode   totalamt  \
0      [{'name': 'Education for all', 'code': '65'}]        65  130000000   
1  [{'name': 'Other economic management', 'code':...     54,24          0   

  totalcommamt                                                url  
0    130000000  http://www.worldbank.org/projects/P129828/ethi...  
1      4700000  http://www.worldbank.org/projects/P144674?lang=en  

[2 rows x 50 columns]

****
## JSON exercise

Using data in file 'data/world_bank_projects.json' and the techniques demonstrated above,
1. Find the 10 countries with most projects
2. Find the top 10 major project themes (using column 'mjtheme_namecode')
3. In 2. above you will notice that some entries have only the code and the name is missing. Create a dataframe with the missing names filled in.

In [15]:
#load in world bank data as string and as data frame
pd.set_option('display.max_columns', 500)

w = json.load((open('data/world_bank_projects.json')))
df = pd.read_json('data/world_bank_projects.json')
df.head()

_id  approvalfy board_approval_month  \
0  {'$oid': '52b213b38594d8a2be17c780'}        1999             November   
1  {'$oid': '52b213b38594d8a2be17c781'}        2015             November   
2  {'$oid': '52b213b38594d8a2be17c782'}        2014             November   
3  {'$oid': '52b213b38594d8a2be17c783'}        2014              October   
4  {'$oid': '52b213b38594d8a2be17c784'}        2014              October   

      boardapprovaldate                                 borrower  \
0  2013-11-12T00:00:00Z  FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA   
1  2013-11-04T00:00:00Z                    GOVERNMENT OF TUNISIA   
2  2013-11-01T00:00:00Z   MINISTRY OF FINANCE AND ECONOMIC DEVEL   
3  2013-10-31T00:00:00Z   MIN. OF PLANNING AND INT'L COOPERATION   
4  2013-10-31T00:00:00Z                      MINISTRY OF FINANCE   

            closingdate                              country_namecode  \
0  2018-07-07T00:00:00Z  Federal Democratic Republic of Ethiopia!$!ET   
1                   NaN                      Republic of Tunisia!$!TN   
2                   NaN                                   Tuvalu!$!TV   
3                   NaN                        Republic of Yemen!$!RY   
4  2019-04-30T00:00:00Z                       Kingdom of Lesotho!$!LS   

  countrycode                              countryname    countryshortname  \
0          ET  Federal Democratic Republic of Ethiopia            Ethiopia   
1          TN                      Republic of Tunisia             Tunisia   
2          TV                                   Tuvalu              Tuvalu   
3          RY                        Republic of Yemen  Yemen, Republic of   
4          LS                       Kingdom of Lesotho             Lesotho   

                                               docty envassesmentcategorycode  \
0  Project Information Document,Indigenous People...                        C   
1  Project Information Document,Integrated Safegu...                        C   
2  Resettlement Plan,Environmental Assessment,Int...                        B   
3  Procurement Plan,Project Information Document,...                        C   
4  Project Information Document,Integrated Safegu...                        B   

   grantamt  ibrdcommamt       id  idacommamt  \
0         0            0  P129828   130000000   
1   4700000            0  P144674           0   
2         0            0  P145310     6060000   
3   1500000            0  P144665           0   
4         0            0  P144933    13100000   

                                  impagency                  lendinginstr  \
0                     MINISTRY OF EDUCATION  Investment Project Financing   
1                       MINISTRY OF FINANCE      Specific Investment Loan   
2  MINISTRY OF TRANSPORT AND COMMUNICATIONS  Investment Project Financing   
3  LABOR INTENSIVE PUBLIC WORKS PROJECT PMU     Technical Assistance Loan   
4            MINISTRY OF TRADE AND INDUSTRY  Investment Project Financing   

  lendinginstrtype  lendprojectcost  \
0               IN        550000000   
1               IN          5700000   
2               IN          6060000   
3               IN          1500000   
4               IN         15000000   

                                 majorsector_percent  \
0  [{'Percent': 46, 'Name': 'Education'}, {'Perce...   
1  [{'Percent': 70, 'Name': 'Public Administratio...   
2       [{'Percent': 100, 'Name': 'Transportation'}]   
3  [{'Percent': 100, 'Name': 'Health and other so...   
4  [{'Percent': 50, 'Name': 'Industry and trade'}...   

                                   mjsector_namecode  \
0  [{'code': 'EX', 'name': 'Education'}, {'code':...   
1  [{'code': 'BX', 'name': 'Public Administration...   
2         [{'code': 'TX', 'name': 'Transportation'}]   
3  [{'code': 'JX', 'name': 'Health and other soci...   
4  [{'code': 'YX', 'name': 'Industry and trade'},...   

                                             mjtheme  \
0                                [Human development]

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 50 columns):
_id                         500 non-null object
approvalfy                  500 non-null int64
board_approval_month        500 non-null object
boardapprovaldate           500 non-null object
borrower                    485 non-null object
closingdate                 370 non-null object
country_namecode            500 non-null object
countrycode                 500 non-null object
countryname                 500 non-null object
countryshortname            500 non-null object
docty                       446 non-null object
envassesmentcategorycode    430 non-null object
grantamt                    500 non-null int64
ibrdcommamt                 500 non-null int64
id                          500 non-null object
idacommamt                  500 non-null int64
impagency                   472 non-null object
lendinginstr                495 non-null object
lendinginstrtype            495 non

1) Find the top 10 countries with the most projects.

In [26]:
pd.set_option('display.max_rows', 500)
df.groupby('countryshortname')['countryshortname'].count().sort_values(ascending = False).head(10)

countryshortname
Indonesia             19
China                 19
Vietnam               17
India                 16
Yemen, Republic of    13
Nepal                 12
Bangladesh            12
Morocco               12
Mozambique            11
Africa                11
Name: countryshortname, dtype: int64

2) Find the top 10 major project themes (using column 'mjtheme_namecode')

In [51]:
#Inefficient way unpacking each row of columns to get project theme name
theme_list = []
for i in df['mjtheme_namecode']:
    for dic in i:
        for k,v in dic.items():
            if k == 'name':
                theme_list.append(v)

In [55]:
#Convert list to Pandas series
theme_df = pd.Series(theme_list)

In [57]:
theme_df.value_counts()


Environment and natural resources management    223
Rural development                               202
Human development                               197
Public sector governance                        184
Social protection and risk management           158
Financial and private sector development        130
                                                122
Social dev/gender/inclusion                     119
Trade and integration                            72
Urban development                                47
Economic management                              33
Rule of law                                      12
dtype: int64

In [68]:
#Use normalization method instead

theme = json_normalize(w, 'mjtheme_namecode')['name']
theme = pd.Series(theme)
theme.value_counts()


Environment and natural resources management    223
Rural development                               202
Human development                               197
Public sector governance                        184
Social protection and risk management           158
Financial and private sector development        130
                                                122
Social dev/gender/inclusion                     119
Trade and integration                            72
Urban development                                47
Economic management                              33
Rule of law                                      12
Name: name, dtype: int64

3) In 2. above you will notice that some entries have only the code and the name is missing. Create a dataframe with the missing names filled in.

There are 122 entries with missing project themes. 2 approaches will be used to impute the missing values. The first is to simply replace the empty entries with 'Missing Name'. The second approach is to match the name of an entry with the same code as that of the missing entry. 

In [69]:
theme = pd.DataFrame(theme)

In [73]:
#Replace missing entries
theme['name'] = theme['name'].replace('','Missing Name')

In [74]:
#Verify all blank entries have been replaced with 'Missing Name'
theme['name'].value_counts()

Environment and natural resources management    223
Rural development                               202
Human development                               197
Public sector governance                        184
Social protection and risk management           158
Financial and private sector development        130
Missing Name                                    122
Social dev/gender/inclusion                     119
Trade and integration                            72
Urban development                                47
Economic management                              33
Rule of law                                      12
Name: name, dtype: int64

In [150]:
#Normalize again

theme_norm = json_normalize(w, 'mjtheme_namecode')
theme_norm = pd.DataFrame(theme_norm)
theme_norm['code'] = theme_norm['code'].astype('int64')

theme_norm['name'].unique()

array(['Human development', '', 'Economic management',
       'Social protection and risk management', 'Trade and integration',
       'Public sector governance',
       'Environment and natural resources management',
       'Social dev/gender/inclusion',
       'Financial and private sector development', 'Rural development',
       'Urban development', 'Rule of law'], dtype=object)

In [151]:
#Create mapping dictionary with code/name combinations

codename_dict = {1: 'Economic management', 2: 'Public sector governance', 3: 'Rule of law', 4: 'Financial and private sector development',
                5: 'Trade and integration', 6: 'Social protection and risk management', 7: 'Social dev/gender/inclusion',
                8: 'Human development', 9: 'Urban development', 10: 'Rural development', 11: 'Environment and natural resources management'}

codename_dict

{1: 'Economic management',
 2: 'Public sector governance',
 3: 'Rule of law',
 4: 'Financial and private sector development',
 5: 'Trade and integration',
 6: 'Social protection and risk management',
 7: 'Social dev/gender/inclusion',
 8: 'Human development',
 9: 'Urban development',
 10: 'Rural development',
 11: 'Environment and natural resources management'}

In [153]:
theme_norm.drop(['name'], axis = 1, inplace = True)
theme_norm

code
0        8
1       11
2        1
3        6
4        5
5        2
6       11
7        6
8        7
9        7
10       5
11       4
12       6
13       6
14       2
15       4
16      11
17       8
18      10
19       7
20       2
21       2
22       2
23      10
24       2
25      10
26       6
27       6
28      11
29       4
30       5
31      11
32      10
33       6
34       6
35       6
36      10
37      11
38       5
39      10
40       2
41       8
42       7
43       8
44       8
45       2
46       7
47       8
48       2
49       2
50       2
51       7
52       7
53       9
54       7
55       7
56       7
57       8
58       8
59      10
60      10
61      10
62       6
63      11
64      11
65      11
66       4
67       5
68       2
69       8
70      11
71       2
72       5
73       6
74      10
75      10
76       7
77       7
78      11
79       8
80       4
81      11
82      11
83      11
84      11
85      11
86       4
87       6
88       1
89       4
90       2
91       8
92       5
93       5
94       7
95      10
96       4
97      10
98      11
99       6
100      7
101     11
102     11
103      6
104     11
105     11
106      2
107      4
108     11
109      6
110     11
111     11
112      4
113      8
114      2
115      2
116      2
117      8
118     11
119      7
120      8
121     10
122      2
123     10
124      4
125      4
126     11
127     11
128      4
129      4
130      7
131      6
132      2
133     11
134      4
135     11
136     11
137     11
138     11
139     11
140     11
141     11
142     10
143      2
144      5
145      2
146      8
147      8
148      8
149     11
150      7
151     11
152     11
153     11
154      4
155      4
156     11
157      2
158     10
159     10
160     11
161     11
162      2
163      7
164      2
165     10
166      5
167      6
168     10
169     10
170     11
171      4
172      6
173      2
174      6
175      1
176      2
177     11
178      2
179      4
180      7
181     10
182      5
183      9
184     11
185      7
186      8
187     11
188      7
189      6
190      8
191      8
192      8
193      4
194      9
195     10
196     11
197      4
198      4
199      4
200      9
201      5
202      2
203      5
204      1
205      1
206      8
207      4
208      4
209     11
210     11
211      2
212      1
213      6
214      6
215      6
216      2
217     10
218      2
219      4
220      1
221      6
222      1
223      1
224      7
225      5
226      2
227      2
228      8
229      8
230      4
231      2
232      6
233     11
234      2
235      4
236      7
237      8
238      8
239      8
240      8
241      4
242     11
243     11
244      4
245      8
246      5
247      4
248      7
249      1
...    ...
1249    10
1250    10
1251    10
1252     6
1253     4
1254     8
1255     8
1256     6
1257     1
1258     8
1259     6
1260     1
1261     7
1262     8
1263     4
1264    10
1265     5
1266     5
1267    10
1268     2
1269     7
1270     7
1271    10
1272     8
1273    10
1274     2
1275     2
1276     4
1277     4
1278     8
1279     4
1280     7
1281     2
1282     2
1283    11
1284    11
1285     7
1286    11
1287     3
1288     4
1289     7
1290     9
1291     2
1292     4
1293     2
1294     4
1295     2
1296    10
1297     7
1298     7
1299    10
1300    10
1301     7
1302     3
1303     9
1304     8
1305     8
1306     8
1307     8
1308     8
1309    10
1310    10
1311    10
1312     4
1313     8
1314    10
1315    10
1316     3
1317     7
1318     9
1319     5
1320     8
1321    11
1322     8
1323     7
1324    11
1325    11
1326     8
1327     7
1328     7
1329     4
1330     8
1331    11
1332    10
1333    10
1334     7
1335    10
1336    10
1337     4
1338     2
1339     4
1340     6
1341     6
1342     9
1343    11
1344    10
1345     5
1346     7
1347     9
1348     8
1349     7
1350     8
1351     2
1352    10
1353     2
1354     7
1355     8
1356     2
1357     8
1358     2
1359     8
1360    11
13

In [154]:
#Replace missing name values with mapped dictionary values
theme_norm['name'] = theme_norm['code'].replace(codename_dict)

In [157]:
#There are no longer any missing/blank values
theme_norm['name'].value_counts()

Environment and natural resources management    250
Rural development                               216
Human development                               210
Public sector governance                        199
Social protection and risk management           168
Financial and private sector development        146
Social dev/gender/inclusion                     130
Trade and integration                            77
Urban development                                50
Economic management                              38
Rule of law                                      15
Name: name, dtype: int64

In [158]:
theme_norm.head()

code                                          name
0     8                             Human development
1    11  Environment and natural resources management
2     1                           Economic management
3     6         Social protection and risk management
4     5                         Trade and integration